src/MongoDB.Driver.Core/Core/Authentication/MongoAWSAuthenticator.cs

```diff
@@ -0,0 +1,531 @@
+            string username,
+            IEnumerable<KeyValuePair<string, string>> properties,
+            IRandomByteGenerator randomByteGenerator,
+            IClock dateTimeProvider)
+            : base(CreateMechanism(username, null, properties, randomByteGenerator, dateTimeProvider))
+        {
+        }
+
+        /// <inheritdoc/>
+        public override string DatabaseName
+        {
+            get { return "$external"; }
+        }
+
+        // private static methods
+        private static MongoAWSMechanism CreateMechanism(
+            UsernamePasswordCredential credential,
+            IEnumerable<KeyValuePair<string, string>> properties,
+            IRandomByteGenerator randomByteGenerator,
+            IClock dateTimeProvider)
+        {
+            if (credential.Source != "$external")
+            {
+                throw new ArgumentException("MONGO AWS authentication may only use the $external source.", nameof(credential));
+            }
+
+            return CreateMechanism(credential.Username, credential.Password, properties, randomByteGenerator, dateTimeProvider);
+        }
+
+        private static MongoAWSMechanism CreateMechanism(
+            string username,
+            SecureString securePassword,
+            IEnumerable<KeyValuePair<string, string>> properties,
+            IRandomByteGenerator randomByteGenerator,
+            IClock dateTimeProvider)
+        {
+            var relativeUri = Environment.GetEnvironmentVariable("AWS_CONTAINER_CREDENTIALS_RELATIVE_URI");
+            var awsCredentialsCreators = new Func<AwsCredentials>[]
+            {
+                () => CreateAwsCredentialsFromMongoCredentials(username, securePassword, properties),
+                () => CreateAwsCredentialsFromEnvironmentVariables(),
+                () => CreateAwsCredentialsFromEcsResponse(relativeUri),
+                () => CreateAwsCredentialsFromEc2Response()
+            };
+
+            foreach (var awsCredentialsCreator in awsCredentialsCreators)
+            {
+                var awsCredentials = awsCredentialsCreator();
+
+                if (awsCredentials == null)
+                {
+                    continue;
+                }
+
+                ValidateAwsCredentials(awsCredentials);
+
+                if (awsCredentials.Username == null)
+                {
+                    continue;
+                }
+
+                var credentials = new UsernamePasswordCredential("$external", awsCredentials.Username, awsCredentials.Password);
+
+                return new MongoAWSMechanism(credentials, awsCredentials.SessionToken, randomByteGenerator, dateTimeProvider);
+            }
+
+            throw new ArgumentException("A MONGODB-AWS must have access key ID.");
+        }
+
+        private static AwsCredentials CreateAwsCredentialsFromMongoCredentials(string username, SecureString securePassword, IEnumerable<KeyValuePair<string, string>> properties)
+        {
```

Comment:
Should probably add:

    if (username == null)
    {
        return null;
    }

Either return a fully usable credential or return null.